File: editors/clion/src/main/kotlin/com/cmod/intellij/binary/BinaryManager.kt
package com.cmod.intellij.binary

import com.cmod.intellij.settings.CmodSettingsState
import com.intellij.notification.NotificationGroupManager
import com.intellij.notification.NotificationType
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.PathManager
import com.intellij.openapi.components.Service
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.util.SystemInfo
import java.io.*
import java.net.HttpURLConnection
import java.net.URI
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.security.MessageDigest
import java.util.zip.GZIPInputStream
import java.util.zip.ZipInputStream

/**
 * Manages downloading, verifying, and locating the cmod binary.
 *
 * On first use (or version mismatch), downloads the platform-appropriate
 * binary from GitHub Releases, verifies its SHA-256 checksum, and
 * extracts it to the plugin's data directory.
 */
@Service(Service.Level.APP)
class BinaryManager {

    companion object {
        private val LOG = Logger.getInstance(BinaryManager::class.java)
        private const val GITHUB_REPO = "satishbabariya/cmod"
        private const val EXPECTED_VERSION = "0.1.0"

        fun getInstance(): BinaryManager {
            return ApplicationManager.getApplication().getService(BinaryManager::class.java)
        }
    }

    private val binaryName = if (SystemInfo.isWindows) "cmod.exe" else "cmod"
    private val binDir: Path = Paths.get(PathManager.getPluginDataPath(), "cmod", "bin")

    /**
     * Returns the path to the cmod binary, downloading if necessary.
     *
     * Resolution order:
     * 1. User-configured path in settings
     * 2. Previously downloaded binary in plugin data directory
     * 3. System PATH / common locations
     * 4. Auto-download from GitHub Releases
     */
    fun ensureBinary(): String? {
        // 1. User-configured path
        val settings = CmodSettingsState.getInstance()
        if (settings.cmodBinaryPath.isNotBlank()) {
            val configured = File(settings.cmodBinaryPath)
            if (configured.exists() && configured.canExecute()) {
                LOG.info("Using configured cmod binary: ${configured.absolutePath}")
                return configured.absolutePath
            }
            LOG.warn("Configured cmod path not found: ${settings.cmodBinaryPath}")
        }

        // 2. Previously downloaded binary
        val managedPath = binDir.resolve(binaryName)
        if (Files.exists(managedPath) && Files.isExecutable(managedPath)) {
            val version = getVersion(managedPath.toString())
            if (version != null && version.contains(EXPECTED_VERSION)) {
                LOG.info("Using managed cmod binary: $managedPath ($version)")
                return managedPath.toString()
            }
            LOG.info("Managed binary version mismatch (got $version, want $EXPECTED_VERSION)")
        }

        // 3. System PATH and common locations
        val systemBinary = findOnSystem()
        if (systemBinary != null) {
            val version = getVersion(systemBinary)
            if (version != null && version.contains(EXPECTED_VERSION)) {
                LOG.info("Using system cmod binary: $systemBinary ($version)")
                return systemBinary
            }
            LOG.info("System cmod version mismatch (got $version, want $EXPECTED_VERSION)")
        }

        // 4. Auto-download
        val platformInfo = getPlatformInfo()
        if (platformInfo == null) {
            LOG.warn("Unsupported platform: ${SystemInfo.OS_NAME} / ${SystemInfo.OS_ARCH}")
            notify(
                "Unsupported platform for cmod auto-download. Please install cmod manually.",
                NotificationType.ERROR
            )
            return systemBinary
        }

        return downloadBinary(platformInfo)
    }

    private fun downloadBinary(platformInfo: PlatformInfo): String? {
        var result: String? = null

        ProgressManager.getInstance().run(object : Task.WithResult<String?, Exception>(
            null,
            "Downloading cmod v$EXPECTED_VERSION",
            true
        ) {
            override fun compute(indicator: ProgressIndicator): String? {
                indicator.isIndeterminate = false
                indicator.text = "Downloading cmod v$EXPECTED_VERSION..."

                val version = "v$EXPECTED_VERSION"
                val archiveName = "cmod-$version-${platformInfo.target}.${platformInfo.archiveExt}"
                val archiveUrl = "https://github.com/$GITHUB_REPO/releases/download/$version/$archiveName"
                val checksumsUrl = "https://github.com/$GITHUB_REPO/releases/download/$version/checksums-$version.sha256"

                try {
                    // Download archive
                    LOG.info("Downloading $archiveUrl")
                    val archiveBytes = downloadUrl(archiveUrl, indicator)

                    if (indicator.isCanceled) return null

                    // Download and verify checksum
                    indicator.text = "Verifying checksum..."
                    indicator.fraction = 0.8
                    try {
                        val checksumsBytes = downloadUrl(checksumsUrl, null)
                        val checksumsText = String(checksumsBytes, Charsets.UTF_8)
                        val expectedHash = parseChecksumForFile(checksumsText, archiveName)

                        if (expectedHash != null) {
                            val digest = MessageDigest.getInstance("SHA-256")
                            val actualHash = digest.digest(archiveBytes).joinToString("") { "%02x".format(it) }
                            if (actualHash != expectedHash) {
                                LOG.error("Checksum mismatch: expected=$expectedHash, actual=$actualHash")
                                notify("cmod download failed: checksum mismatch", NotificationType.ERROR)
                                return null
                            }
                            LOG.info("Checksum verified: $actualHash")
                        }
                    } catch (e: Exception) {
                        LOG.warn("Could not verify checksum: ${e.message}")
                    }

                    // Extract binary
                    indicator.text = "Extracting binary..."
                    indicator.fraction = 0.9
                    Files.createDirectories(binDir)

                    val destPath = binDir.resolve(binaryName)
                    if (platformInfo.archiveExt == "tar.gz") {
                        extractTarGz(archiveBytes, destPath.toString(), binaryName)
                    } else {
                        extractZip(archiveBytes, destPath.toString(), binaryName)
                    }

                    // Set executable permission on Unix
                    if (!SystemInfo.isWindows) {
                        destPath.toFile().setExecutable(true, false)
                    }

                    indicator.fraction = 1.0
                    LOG.info("cmod binary installed to $destPath")
                    notify("cmod v$EXPECTED_VERSION downloaded successfully.", NotificationType.INFORMATION)

                    return destPath.toString()
                } catch (e: Exception) {
                    LOG.error("Failed to download cmod binary", e)
                    notify(
                        "Failed to download cmod: ${e.message}. Please install manually.",
                        NotificationType.ERROR
                    )
                    return null
                }
            }
        }.also { result = it.compute(ProgressManager.getInstance().progressIndicator ?: EmptyProgressIndicator()) })

        return result
    }

    private fun downloadUrl(url: String, indicator: ProgressIndicator?): ByteArray {
        var currentUrl = url
        var redirectCount = 0
        while (redirectCount < 10) {
            val connection = URI(currentUrl).toURL().openConnection() as HttpURLConnection
            connection.setRequestProperty("User-Agent", "cmod-clion")
            connection.instanceFollowRedirects = false
            connection.connect()

            val responseCode = connection.responseCode
            if (responseCode in 300..399) {
                currentUrl = connection.getHeaderField("Location")
                    ?: throw IOException("Redirect with no Location header")
                redirectCount++
                connection.disconnect()
                continue
            }

            if (responseCode != 200) {
                connection.disconnect()
                throw IOException("HTTP $responseCode for $currentUrl")
            }

            val totalBytes = connection.contentLengthLong
            val output = ByteArrayOutputStream()
            connection.inputStream.use { input ->
                val buffer = ByteArray(8192)
                var bytesRead: Int
                var totalRead = 0L
                while (input.read(buffer).also { bytesRead = it } != -1) {
                    output.write(buffer, 0, bytesRead)
                    totalRead += bytesRead
                    if (indicator != null && totalBytes > 0) {
                        indicator.fraction = totalRead.toDouble() / totalBytes * 0.8
                    }
                }
            }
            connection.disconnect()
            return output.toByteArray()
        }
        throw IOException("Too many redirects for $url")
    }

    private fun extractTarGz(archiveBytes: ByteArray, destPath: String, binaryName: String) {
        val gzipInput = GZIPInputStream(ByteArrayInputStream(archiveBytes))
        val tarBytes = gzipInput.readBytes()
        gzipInput.close()

        // Simple tar parser: 512-byte headers
        var offset = 0
        while (offset < tarBytes.size - 512) {
            val header = tarBytes.sliceArray(offset until offset + 512)
            if (header.all { it == 0.toByte() }) break

            val fileName = String(header, 0, 100, Charsets.UTF_8).trimEnd('\u0000').trim()
            val sizeStr = String(header, 124, 12, Charsets.UTF_8).trimEnd('\u0000').trim()
            val fileSize = if (sizeStr.isNotEmpty()) sizeStr.toLong(8) else 0L

            offset += 512 // Past header

            if (fileName == binaryName || fileName.endsWith("/$binaryName")) {
                val fileData = tarBytes.sliceArray(offset until offset + fileSize.toInt())
                File(destPath).writeBytes(fileData)
                return
            }

            offset += (Math.ceil(fileSize.toDouble() / 512) * 512).toInt()
        }
        throw IOException("Binary '$binaryName' not found in tar.gz archive")
    }

    private fun extractZip(archiveBytes: ByteArray, destPath: String, binaryName: String) {
        ZipInputStream(ByteArrayInputStream(archiveBytes)).use { zis ->
            var entry = zis.nextEntry
            while (entry != null) {
                val name = entry.name
                if (name == binaryName || name.endsWith("/$binaryName")) {
                    File(destPath).outputStream().use { out ->
                        zis.copyTo(out)
                    }
                    return
                }
                entry = zis.nextEntry
            }
        }
        throw IOException("Binary '$binaryName' not found in zip archive")
    }

    private fun parseChecksumForFile(checksumsText: String, fileName: String): String? {
        for (line in checksumsText.lines()) {
            val trimmed = line.trim()
            if (trimmed.isEmpty()) continue
            val parts = trimmed.split(Regex("\\s+"))
            if (parts.size >= 2) {
                val hash = parts[0]
                val name = parts.last().removePrefix("*")
                if (name == fileName || name.endsWith("/$fileName")) {
                    return hash
                }
            }
        }
        return null
    }

    private fun getVersion(binaryPath: String): String? {
        return try {
            val process = ProcessBuilder(binaryPath, "--version")
                .redirectErrorStream(true)
                .start()
            val output = process.inputStream.bufferedReader().readText().trim()
            val exitCode = process.waitFor()
            if (exitCode == 0) output else null
        } catch (e: Exception) {
            null
        }
    }

    private fun findOnSystem(): String? {
        // Search PATH
        val pathEnv = System.getenv("PATH") ?: ""
        val separator = if (SystemInfo.isWindows) ";" else ":"
        for (dir in pathEnv.split(separator)) {
            val candidate = Paths.get(dir, binaryName)
            if (Files.exists(candidate) && Files.isExecutable(candidate)) {
                return candidate.toAbsolutePath().toString()
            }
        }

        // Common install locations
        val commonPaths = listOf(
            Paths.get(System.getProperty("user.home"), ".cargo", "bin", binaryName),
            Paths.get(System.getProperty("user.home"), ".local", "bin", binaryName),
            Paths.get("/usr", "local", "bin", binaryName),
            Paths.get("/opt", "homebrew", "bin", binaryName),
        )
        for (p in commonPaths) {
            if (Files.exists(p) && Files.isExecutable(p)) {
                return p.toAbsolutePath().toString()
            }
        }
        return null
    }

    private fun notify(message: String, type: NotificationType) {
        ApplicationManager.getApplication().invokeLater {
            NotificationGroupManager.getInstance()
                .getNotificationGroup("cmod.notifications")
                .createNotification(message, type)
                .notify(null)
        }
    }

    data class PlatformInfo(
        val target: String,
        val archiveExt: String,
    )

    private fun getPlatformInfo(): PlatformInfo? {
        val osName = SystemInfo.OS_NAME.lowercase()
        val arch = System.getProperty("os.arch")?.lowercase() ?: return null

        return when {
            SystemInfo.isLinux && (arch == "amd64" || arch == "x86_64") ->
                PlatformInfo("x86_64-unknown-linux-gnu", "tar.gz")
            SystemInfo.isLinux && (arch == "aarch64" || arch == "arm64") ->
                PlatformInfo("aarch64-unknown-linux-gnu", "tar.gz")
            SystemInfo.isMac && (arch == "amd64" || arch == "x86_64") ->
                PlatformInfo("x86_64-apple-darwin", "tar.gz")
            SystemInfo.isMac && (arch == "aarch64" || arch == "arm64") ->
                PlatformInfo("aarch64-apple-darwin", "tar.gz")
            SystemInfo.isWindows && (arch == "amd64" || arch == "x86_64") ->
                PlatformInfo("x86_64-pc-windows-msvc", "zip")
            SystemInfo.isWindows && (arch == "aarch64" || arch == "arm64") ->
                PlatformInfo("aarch64-pc-windows-msvc", "zip")
            else -> null
        }
    }

    /**
     * Minimal progress indicator for non-UI contexts.
     */
    private class EmptyProgressIndicator : ProgressIndicator {
        override fun start() {}
        override fun stop() {}
        override fun isRunning(): Boolean = true
        override fun cancel() {}
        override fun isCanceled(): Boolean = false
        override fun setText(text: String?) {}
        override fun getText(): String = ""
        override fun setText2(text: String?) {}
        override fun getText2(): String = ""
        override fun getFraction(): Double = 0.0
        override fun setFraction(fraction: Double) {}
        override fun pushState() {}
        override fun popState() {}
        override fun isModal(): Boolean = false
        override fun getModalityState(): com.intellij.openapi.application.ModalityState =
            com.intellij.openapi.application.ModalityState.nonModal()
        override fun setModalityState(modalityState: com.intellij.openapi.application.ModalityState) {}
        override fun isIndeterminate(): Boolean = true
        override fun setIndeterminate(indeterminate: Boolean) {}
        override fun checkCanceled() {}
        override fun isPopupWasShown(): Boolean = false
        override fun isShowing(): Boolean = false
    }
}

File: editors/vscode/src/utils/binaryManager.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as https from 'https';
import * as http from 'http';
import * as crypto from 'crypto';
import * as cp from 'child_process';
import { createGunzip, inflateRawSync } from 'zlib';

/** The cmod version this extension expects. Kept in sync with package.json. */
const EXPECTED_VERSION: string = require('../../package.json').cmod?.binaryVersion ?? require('../../package.json').version;

const GITHUB_REPO = 'satishbabariya/cmod';

interface PlatformInfo {
    target: string;
    binaryName: string;
    archiveExt: string;
}

/**
 * Resolves the current platform to a cmod release target triple.
 */
function getPlatformInfo(): PlatformInfo | undefined {
    const platform = process.platform;
    const arch = process.arch;
    const binaryName = platform === 'win32' ? 'cmod.exe' : 'cmod';

    const mapping: Record<string, Record<string, { target: string; ext: string }>> = {
        linux: {
            x64: { target: 'x86_64-unknown-linux-gnu', ext: 'tar.gz' },
            arm64: { target: 'aarch64-unknown-linux-gnu', ext: 'tar.gz' },
        },
        darwin: {
            x64: { target: 'x86_64-apple-darwin', ext: 'tar.gz' },
            arm64: { target: 'aarch64-apple-darwin', ext: 'tar.gz' },
        },
        win32: {
            x64: { target: 'x86_64-pc-windows-msvc', ext: 'zip' },
            arm64: { target: 'aarch64-pc-windows-msvc', ext: 'zip' },
        },
    };

    // Check for Alpine/musl
    if (platform === 'linux' && arch === 'x64') {
        try {
            const lddOutput = cp.execSync('ldd --version 2>&1 || true', { encoding: 'utf-8' });
            if (lddOutput.includes('musl')) {
                return { target: 'x86_64-unknown-linux-musl', binaryName, archiveExt: 'tar.gz' };
            }
        } catch {
            // Fall through to glibc default
        }
    }

    const entry = mapping[platform]?.[arch];
    if (!entry) {
        return undefined;
    }

    return { target: entry.target, binaryName, archiveExt: entry.ext };
}

/** Default timeout in milliseconds for HTTP requests. */
const REQUEST_TIMEOUT_MS = 30000;
const SOCKET_TIMEOUT_MS = 15000;

/**
 * Downloads a URL, following redirects, and returns the data as a Buffer.
 */
function download(url: string, onProgress?: (percent: number) => void, timeoutMs: number = REQUEST_TIMEOUT_MS): Promise<Buffer> {
    return new Promise((resolve, reject) => {
        const get = url.startsWith('https') ? https.get : http.get;
        const request = get(url, { headers: { 'User-Agent': 'cmod-vscode' } }, (res) => {
            // Follow redirects (pass timeout to recursive call)
            if (res.statusCode && res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
                download(res.headers.location, onProgress, timeoutMs).then(resolve, reject);
                return;
            }

            if (res.statusCode !== 200) {
                reject(new Error(`Download failed: HTTP ${res.statusCode} for ${url}`));
                return;
            }

            const totalBytes = parseInt(res.headers['content-length'] || '0', 10);
            const chunks: Buffer[] = [];
            let receivedBytes = 0;

            res.on('data', (chunk: Buffer) => {
                chunks.push(chunk);
                receivedBytes += chunk.length;
                if (onProgress && totalBytes > 0) {
                    onProgress(Math.round((receivedBytes / totalBytes) * 100));
                }
            });
            res.on('end', () => resolve(Buffer.concat(chunks)));
            res.on('error', reject);
        });

        // Set request timeout
        request.setTimeout(timeoutMs, () => {
            request.destroy();
            reject(new Error(`Download timeout after ${timeoutMs}ms for ${url}`));
        });

        // Set socket timeout when socket is assigned
        request.on('socket', (socket) => {
            socket.setTimeout(SOCKET_TIMEOUT_MS);
            socket.on('timeout', () => {
                request.destroy();
                reject(new Error(`Socket timeout after ${SOCKET_TIMEOUT_MS}ms for ${url}`));
            });
        });

        request.on('error', reject);
    });
}

/**
 * Extracts cmod binary from a tar.gz archive buffer into the target directory.
 */
async function extractTarGz(archiveBuffer: Buffer, destDir: string, binaryName: string): Promise<void> {
    // Simple tar extraction: gunzip then parse tar format
    const gunzipped = await new Promise<Buffer>((resolve, reject) => {
        const gunzip = createGunzip();
        const chunks: Buffer[] = [];
        gunzip.on('data', (chunk: Buffer) => chunks.push(chunk));
        gunzip.on('end', () => resolve(Buffer.concat(chunks)));
        gunzip.on('error', reject);
        gunzip.end(archiveBuffer);
    });

    // Parse tar: find the binary entry
    let offset = 0;
    while (offset < gunzipped.length - 512) {
        const header = gunzipped.subarray(offset, offset + 512);
        // Check for empty block (end of archive)
        if (header.every((b) => b === 0)) {
            break;
        }

        const fileName = header.subarray(0, 100).toString('utf-8').replace(/\0/g, '').trim();
        const sizeOctal = header.subarray(124, 136).toString('utf-8').replace(/\0/g, '').trim();
        const fileSize = parseInt(sizeOctal, 8) || 0;

        offset += 512; // Move past header

        if (fileName === binaryName || fileName.endsWith('/' + binaryName)) {
            const fileData = gunzipped.subarray(offset, offset + fileSize);
            const destPath = path.join(destDir, binaryName);
            fs.writeFileSync(destPath, fileData);
            fs.chmodSync(destPath, 0o755);
            return;
        }

        // Skip to next entry (aligned to 512 bytes)
        offset += Math.ceil(fileSize / 512) * 512;
    }

    throw new Error(`Binary '${binaryName}' not found in archive`);
}

/**
 * Extracts cmod.exe from a zip archive buffer into the target directory.
 *
 * Uses a minimal zip parser — no external dependencies needed.
 */
async function extractZip(archiveBuffer: Buffer, destDir: string, binaryName: string): Promise<void> {
    // Find End of Central Directory record
    let eocdOffset = -1;
    for (let i = archiveBuffer.length - 22; i >= 0; i--) {
        if (archiveBuffer.readUInt32LE(i) === 0x06054b50) {
            eocdOffset = i;
            break;
        }
    }
    if (eocdOffset === -1) {
        throw new Error('Invalid zip archive: EOCD not found');
    }

    const centralDirOffset = archiveBuffer.readUInt32LE(eocdOffset + 16);
    const centralDirEntries = archiveBuffer.readUInt16LE(eocdOffset + 10);

    let cdOffset = centralDirOffset;
    for (let i = 0; i < centralDirEntries; i++) {
        if (archiveBuffer.readUInt32LE(cdOffset) !== 0x02014b50) {
            break;
        }

        const fileNameLen = archiveBuffer.readUInt16LE(cdOffset + 28);
        const extraLen = archiveBuffer.readUInt16LE(cdOffset + 30);
        const commentLen = archiveBuffer.readUInt16LE(cdOffset + 32);
        const localHeaderOffset = archiveBuffer.readUInt32LE(cdOffset + 42);
        const fileName = archiveBuffer.subarray(cdOffset + 46, cdOffset + 46 + fileNameLen).toString('utf-8');

        if (fileName === binaryName || fileName.endsWith('/' + binaryName)) {
            // Read from local file header
            const lfhNameLen = archiveBuffer.readUInt16LE(localHeaderOffset + 26);
            const lfhExtraLen = archiveBuffer.readUInt16LE(localHeaderOffset + 28);
            const compressedSize = archiveBuffer.readUInt32LE(localHeaderOffset + 18);
            const uncompressedSize = archiveBuffer.readUInt32LE(localHeaderOffset + 22);
            const dataOffset = localHeaderOffset + 30 + lfhNameLen + lfhExtraLen;

            const compressionMethod = archiveBuffer.readUInt16LE(localHeaderOffset + 8);
            const compressedData = archiveBuffer.subarray(dataOffset, dataOffset + compressedSize);

            let fileData: Buffer;
            if (compressionMethod === 0) {
                // Stored (no compression)
                fileData = compressedData;
            } else if (compressionMethod === 8) {
                // DEFLATE compression
                try {
                    fileData = inflateRawSync(compressedData);
                    if (fileData.length !== uncompressedSize) {
                        throw new Error(
                            `Decompressed size mismatch: expected ${uncompressedSize}, got ${fileData.length}`
                        );
                    }
                } catch (err) {
                    throw new Error(`Failed to decompress DEFLATE entry: ${err}`);
                }
            } else {
                throw new Error(
                    `Unsupported zip compression method ${compressionMethod}; only stored (0) and DEFLATE (8) are supported`
                );
            }

            const destPath = path.join(destDir, binaryName);
            fs.writeFileSync(destPath, fileData);
            return;
        }

        cdOffset += 46 + fileNameLen + extraLen + commentLen;
    }

    throw new Error(`Binary '${binaryName}' not found in archive`);
}

/**
 * Manages downloading and verifying the cmod binary.
 */
export class BinaryManager {
    private readonly binDir: string;
    private readonly outputChannel: vscode.OutputChannel;

    constructor(context: vscode.ExtensionContext, outputChannel: vscode.OutputChannel) {
        this.binDir = path.join(context.globalStorageUri.fsPath, 'bin');
        this.outputChannel = outputChannel;
    }

    /**
     * Returns the path to the cmod binary, downloading it if necessary.
     *
     * Resolution order:
     * 1. User-configured `cmod.path` setting
     * 2. Bundled binary (platform-specific VSIX)
     * 3. Previously downloaded binary in global storage
     * 4. `cmod` on system PATH
     * 5. Auto-download from GitHub Releases
     */
    async ensureBinary(): Promise<string> {
        // 1. User-configured path
        const config = vscode.workspace.getConfiguration('cmod');
        const configuredPath = config.get<string>('path', '').trim();
        if (configuredPath) {
            const resolved = this.resolveHome(configuredPath);
            if (fs.existsSync(resolved)) {
                this.outputChannel.appendLine(`Using configured cmod binary: ${resolved}`);
                return resolved;
            }
            this.outputChannel.appendLine(`Configured path not found: ${resolved}, trying other sources.`);
        }

        const platformInfo = getPlatformInfo();
        const binaryName = platformInfo?.binaryName ?? (process.platform === 'win32' ? 'cmod.exe' : 'cmod');

        // 2. Bundled binary (shipped inside the VSIX for platform-specific packages)
        const bundledPath = path.join(__dirname, '..', 'bin', binaryName);
        if (fs.existsSync(bundledPath)) {
            this.outputChannel.appendLine(`Using bundled cmod binary: ${bundledPath}`);
            return bundledPath;
        }

        // 3. Previously downloaded binary
        const managedPath = path.join(this.binDir, binaryName);
        if (fs.existsSync(managedPath)) {
            const version = this.getVersion(managedPath);
            if (version && version.includes(EXPECTED_VERSION)) {
                this.outputChannel.appendLine(`Using managed cmod binary: ${managedPath} (${version})`);
                return managedPath;
            }
            this.outputChannel.appendLine(`Managed binary version mismatch (got ${version}, want ${EXPECTED_VERSION}), will re-download.`);
        }

        // 4. System PATH
        const systemPath = this.findOnPath(binaryName);
        if (systemPath) {
            const version = this.getVersion(systemPath);
            if (version && version.includes(EXPECTED_VERSION)) {
                this.outputChannel.appendLine(`Using system cmod binary: ${systemPath} (${version})`);
                return systemPath;
            }
            // System binary exists but wrong version — still try auto-download
            this.outputChannel.appendLine(`System cmod version mismatch (got ${version}, want ${EXPECTED_VERSION}).`);
        }

        // 5. Auto-download
        if (!platformInfo) {
            const msg = `Unsupported platform: ${process.platform}/${process.arch}. Please install cmod manually and set cmod.path.`;
            vscode.window.showErrorMessage(msg);
            // Fall back to system PATH binary or bare name
            return systemPath ?? 'cmod';
        }

        return this.downloadBinary(platformInfo);
    }

    private async downloadBinary(platformInfo: PlatformInfo): Promise<string> {
        const version = `v${EXPECTED_VERSION}`;
        const archiveName = `cmod-${version}-${platformInfo.target}.${platformInfo.archiveExt}`;
        const archiveUrl = `https://github.com/${GITHUB_REPO}/releases/download/${version}/${archiveName}`;
        const checksumsUrl = `https://github.com/${GITHUB_REPO}/releases/download/${version}/checksums-${version}.sha256`;

        const destPath = path.join(this.binDir, platformInfo.binaryName);

        return vscode.window.withProgress(
            {
                location: vscode.ProgressLocation.Notification,
                title: 'cmod',
                cancellable: true,
            },
            async (progress, token) => {
                progress.report({ message: `Downloading cmod ${version}...` });
                this.outputChannel.appendLine(`Downloading ${archiveUrl}`);

                if (token.isCancellationRequested) {
                    throw new Error('Download cancelled');
                }

                // Download archive
                const archiveBuffer = await download(archiveUrl, (percent) => {
                    progress.report({ message: `Downloading cmod ${version}... ${percent}%`, increment: 0 });
                });

                if (token.isCancellationRequested) {
                    throw new Error('Download cancelled');
                }

                // Download and verify checksum
                progress.report({ message: 'Verifying checksum...' });
                try {
                    const checksumsData = await download(checksumsUrl);
                    const checksumsText = checksumsData.toString('utf-8');
                    const expectedHash = this.parseChecksumForFile(checksumsText, archiveName);

                    if (expectedHash) {
                        const actualHash = crypto.createHash('sha256').update(archiveBuffer).digest('hex');
                        if (actualHash !== expectedHash) {
                            throw new Error(
                                `Checksum mismatch for ${archiveName}:\n  expected: ${expectedHash}\n  actual:   ${actualHash}`
                            );
                        }
                        this.outputChannel.appendLine(`Checksum verified: ${actualHash}`);
                    } else {
                        this.outputChannel.appendLine(`Warning: no checksum found for ${archiveName}, skipping verification.`);
                    }
                } catch (err) {
                    if (err instanceof Error && err.message.includes('Checksum mismatch')) {
                        throw err;
                    }
                    this.outputChannel.appendLine(`Warning: could not verify checksum: ${err}`);
                }

                // Extract binary
                progress.report({ message: 'Extracting binary...' });
                fs.mkdirSync(this.binDir, { recursive: true });

                if (platformInfo.archiveExt === 'tar.gz') {
                    await extractTarGz(archiveBuffer, this.binDir, platformInfo.binaryName);
                } else {
                    await extractZip(archiveBuffer, this.binDir, platformInfo.binaryName);
                }

                this.outputChannel.appendLine(`cmod binary installed to ${destPath}`);
                vscode.window.showInformationMessage(`cmod ${version} downloaded successfully.`);
                return destPath;
            }
        );
    }

    private parseChecksumForFile(checksumsText: string, fileName: string): string | undefined {
        for (const line of checksumsText.split('\n')) {
            const trimmed = line.trim();
            if (!trimmed) {
                continue;
            }
            // Format: "hash  filename" or "hash *filename"
            const parts = trimmed.split(/\s+/);
            if (parts.length >= 2) {
                const hash = parts[0];
                const name = parts[parts.length - 1].replace(/^\*/, '');
                if (name === fileName || name.endsWith('/' + fileName)) {
                    return hash;
                }
            }
        }
        return undefined;
    }

    private getVersion(binaryPath: string): string | undefined {
        try {
            const result = cp.execSync(`"${binaryPath}" --version`, {
                timeout: 5000,
                encoding: 'utf-8',
                stdio: ['pipe', 'pipe', 'pipe'],
            });
            return result.trim();
        } catch {
            return undefined;
        }
    }

    private findOnPath(binaryName: string): string | undefined {
        const pathEnv = process.env.PATH || '';
        const separator = process.platform === 'win32' ? ';' : ':';
        const extensions = process.platform === 'win32' ? ['.exe', '.cmd', '.bat', ''] : [''];

        for (const dir of pathEnv.split(separator)) {
            for (const ext of extensions) {
                const fullPath = path.join(dir, binaryName + (ext && !binaryName.endsWith(ext) ? ext : ''));
                try {
                    fs.accessSync(fullPath, fs.constants.X_OK);
                    return fullPath;
                } catch {
                    // continue
                }
            }
        }
        return undefined;
    }

    private resolveHome(filepath: string): string {
        if (filepath.startsWith('~/') || filepath === '~') {
            const home = process.env.HOME || process.env.USERPROFILE || '';
            return path.join(home, filepath.slice(2));
        }
        return filepath;
    }
}
